compiler/src/dmd/errorsink.d

```diff
@@ -0,0 +1,121 @@
+/**
+ * Provides an abstraction for what to do with error messages.
+ *
+ * Copyright:   Copyright (C) 2023 by The D Language Foundation, All Rights Reserved
+ * Authors:     $(LINK2 https://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 https://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/errorsink.d, _errorsink.d)
+ * Documentation:  https://dlang.org/phobos/dmd_errorsink.html
+ * Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/errorsink.d
+ */
+
+module dmd.errorsink;
+
+import dmd.location;
+
+/***************************************
+ * Where error/warning/deprecation messages go.
+ */
+abstract class ErrorSink
```

Comment:
Reuse [existing wheels](https://github.com/dlang/dmd/blob/aa19d8a145afeb627a1136bf58f958998cd0fcec/compiler/src/dmd/frontend.d#L471-L683) or explain why they are insufficient. 

Such deficiencies might include `dmd.frontend` imports `std`.

Note also that `dmd.frontend` is a stable externally available API and so at worst this PR should move the definition from there to here, and have a publicly imported alias to the new place.

Reply:
> Reuse existing wheels or explain why they are insufficient.

This line illustrates the insufficiency:
```
Returns: false if the message should also be printed to stderr, true otherwise
```
This is the wrong abstraction. The user of Lexer should be able to send the errors anywhere, baking into the API "stderr" or not is insufficent. `ErrorSink` puts no constraint on where the messages should go. They can even go nowhere, as `ErrorSinkNull` demonstrates.

It's other insufficiency is it is centered around a pointer to a function which overrides the default behavior. This is a primitive form of OOP that fell out of favor in the 1980s. The abstract class, for which the user adds derived classes with specific behaviors, is the more modern approach. The abstract class designer does not have to anticipate what the implementation classes will be doing.

Reply:
I don't want to touch dmd.frontend in this PR. As I mentioned to @ibuclaw it needs a significant overhaul.

Reply:
Then I consider this PR to require approval from @jacob-carlborg  before merging.

> don't want to touch dmd.frontend in this PR

I do not trust you will ever get around to this. And this adds yet more files without packaging things properly and I trust even less that you will ever do that.

> They can even go nowhere, as ErrorSinkNull demonstrates.

So can `DiagnositReporter`, `return true;`.

> As I mentioned to @ibuclaw it needs a significant overhaul.

Where? why? (Hint, no it doesn't).